### PR TITLE
[Intg]: Leonardo AI Integration

### LINKED ISSUE #8528: leonardo ai handler
### Description
Integration for leonardo ai image gen - https://app.leonardo.ai/

### Impact
This issue affects users who want to integrate leonardo ai image generation into their projects. Without a proper handler, users will not be able to utilize the leonardo ai image generation capabilities.

### PR DESCRIPTION
**Pull request by tcchawla**
## Description

This PR includes a new integration with Leonardo AI.

Fixes #8528 

## Type of change

(Please delete options that are not relevant)

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ⚡ New feature (non-breaking change which adds functionality)
- [ ] 📢 Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] 📄 This change requires a documentation update

## Verification Process

To ensure the changes are working as expected:

 - [ ]   Test Location: Specify the URL or path for testing.
 - [x]   Verification Steps: Outline the steps or queries needed to validate the change. Include any data, configurations, or actions required to reproduce or see the new functionality.
 
Verification Queries
 ~~~~sql
CREATE ML_ENGINE leo_engine
FROM leonardo_ai
USING
    api_key = '6015572d-11fa-445d-8d62-3b9ad7597e26';


CREATE MODEL mindsdb.leo
PREDICT url
USING
   engine = 'leo_engine',
   model = '6bef9f1b-29cb-40c7-b9df-32b51c1f67d3',
   prompt_template = '{{text}}, 8K | highly detailed realistic 3d oil painting style cyberpunk by MAD DOG JONES combined with Van Gogh  |  cinematic lighting | happy colors';

SELECT *
FROM mindsdb.leo
WHERE text = 'Generate a random ANIME picture';

DESCRIBE MODEL mindsdb.leo;

DESCRIBE MODEL mindsdb.leo.features;

DESCRIBE MODEL mindsdb.leo.metadata;
~~~~

## Additional Media:

- [x] I have attached a brief loom video or screenshots showcasing the new functionality or change.

[Loom Video Link](https://www.loom.com/share/ef8416dddf3d498a8c0ba95653219ea2?sid=24142830-2c27-41e1-92eb-95b9aa009097)

## Checklist:

- [x] My code follows the style guidelines(PEP 8) of MindsDB.
- [x] I have appropriately commented on my code, especially in complex areas.
- [x] Necessary documentation updates are either made or tracked in issues.
- [ ] Relevant unit and integration tests are updated or added.

### CODE REVIEW

**Comment by tcchawla** [comment]
Hey @ZoranPandovski @dusvyat, please review this PR.

**Comment by dusvyat** [comment]
@tcchawla pls include some unit tests.

**Comment by tcchawla** [comment]
@dusvyat, I have added unit tests, please verify.

File: mindsdb/integrations/handlers/leonardoai_handler/leonardo_ai_handler.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+class LeonardoAIHandler(BaseMLEngine):
+    """
+    This integration seamlessly combines MindsDB and Leonardo AI to create a powerful
+    AI-driven solution for creative content generation.
+
+    Content Generation with Leonardo AI: Harness the power of advanced generative
+    models for creative content production. From realistic images to artistic text, Leonardo
+    AI opens up new possibilities for content creators.
+    """
+    name = "leonardoai"
```
Comment (ZoranPandovski) [inline mindsdb/integrations/handlers/leonardoai_handler/leonardo_ai_handler.py:28]: In the README we are referencing `leonardo_ai`. So maybe change this or the README info

File: mindsdb/integrations/handlers/leonardoai_handler/leonardo_ai_handler.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        # Wait for 15 seconds
+
+        # Set the desired duration in seconds
+        duration = 15
+
+        # Record the start time
+        start_time = time.time()
+
+        # Run a busy loop for the specified duration
+        while time.time() - start_time < duration:
```
Comment (ZoranPandovski) [inline mindsdb/integrations/handlers/leonardoai_handler/leonardo_ai_handler.py:193]: Why do we need this? Is there a way to remove it or simplify this?

File: mindsdb/integrations/handlers/leonardoai_handler/leonardo_ai_handler.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+            available_models = self._get_platform_model(args)
+
+            if not args['using']['model']:
+                args['using']['model'] = self.default_model
+            elif args['using']['model'] not in available_models:    # if invalid model_id is provided
+                raise Exception("Invalid Model ID. Please use a valid Model")
+
+        # API key is invalid
+        else:
+            raise Exception("Unable to make connection, please verify the API key.")
```
Comment (ZoranPandovski) [inline mindsdb/integrations/handlers/leonardoai_handler/leonardo_ai_handler.py:71]: Maybe wrap the above in try/except and return Auth exception

File: mindsdb/integrations/handlers/leonardoai_handler/leonardo_ai_handler.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+            "accept": "application/json",
+            "authorization": f"Bearer {api_key}"
+        }
+
+        # payload
+        generation_payload = {
+            "height": 512,
+            "modelId": args['using']['model'],
+            "prompt": f"{prompts}",
+            "width": 512
```
Comment (ZoranPandovski) [inline mindsdb/integrations/handlers/leonardoai_handler/leonardo_ai_handler.py:177]: Should we add width and height as params in USING clause so users can specify them?

**Review by ZoranPandovski — CHANGES_REQUESTED** [review]

File: mindsdb/integrations/handlers/leonardoai_handler/leonardo_ai_handler.py (hunk 1 above)
Comment (tcchawla) [reply mindsdb/integrations/handlers/leonardoai_handler/leonardo_ai_handler.py]: I changed it to `leonardo_ai` in the handler file to make it consistent with the README.

File: mindsdb/integrations/handlers/leonardoai_handler/leonardo_ai_handler.py (hunk 3 above)
Comment (tcchawla) [reply mindsdb/integrations/handlers/leonardoai_handler/leonardo_ai_handler.py]: I wrapped the if/else within try/except block.

File: mindsdb/integrations/handlers/leonardoai_handler/leonardo_ai_handler.py (hunk 4 above)
Comment (tcchawla) [reply mindsdb/integrations/handlers/leonardoai_handler/leonardo_ai_handler.py]: User can now pass height and width, if user doesn't provide it, default values will be used.

**Comment by tcchawla** [comment]
Hey @ZoranPandovski, I have made the necessary changes based on the review comments.

**Comment by ZoranPandovski** [comment]
Thanks. Let's wait for a quick review from @paxcema and we can merge this. Thank you @tcchawla

**Review by dusvyat — APPROVED** [review]
Thanks for your changes, LGTM!

**Pull request merged**